src/homebridge-roku.js

```diff
@@ -87,74 +164,76 @@ class RokuAccessory {
           .volumeUp()
           .exec(cmd => this.muted && cmd.volumeMute())
           .send()
-          .then(() => callback(null))
-          .catch(callback);
+          .then(() => callback(null), callback);
       });
 
-    return volume;
-  }
-
-  setupVolumeUp() {
-    return this.setupVolume(keys.VOLUME_UP, this.volumeIncrement);
-  }
-
-  setupVolumeDown() {
-    return this.setupVolume(keys.VOLUME_DOWN, this.volumeDecrement);
-  }
-
-  setupVolume(key, increment) {
-    const volume = new Service.Switch(
-      `${this.name} ${key.command}`,
-      key.command,
-    );
-
-    volume
-      .getCharacteristic(Characteristic.On)
-      .on('get', callback => callback(null, false))
-      .on('set', (value, callback) => {
-        this.roku
-          .command()
-          .keypress(key, increment)
-          .send()
-          .then(() => callback(null, false))
-          .catch(callback);
-      });
-
-    return volume;
-  }
-
-  setupChannels() {
-    return map(this.appMap, (id, name) => this.setupChannel(name, id));
-  }
-
-  setupChannel(name, id) {
-    const channel = new Service.Switch(`${this.name} ${name}`, name);
-
-    channel
-      .getCharacteristic(Characteristic.On)
-      .on('get', callback => {
-        this.roku
-          .active()
-          .then(app => {
-            callback(null, app && app.id === id);
-          })
-          .catch(callback);
-      })
-      .on('set', (value, callback) => {
-        if (value) {
+    speaker
+      .getCharacteristic(Characteristic.VolumeSelector)
+      .on('set', (newValue, callback) => {
+        if (newValue == Characteristic.VolumeSelector.INCREMENT) {
           this.roku
-            .launch(id)
-            .then(() => callback(null, true))
-            .catch(callback);
+            .command()
+            .keypress(keys.VOLUME_UP, this.volumeIncrement)
+            .send()
+            .then(() => callback(null), callback);
         } else {
           this.roku
-            .keypress(keys.HOME)
-            .then(() => callback(null, false))
-            .catch(callback);
+            .command()
+            .keypress(keys.VOLUME_DOWN, this.volumeDecrement)
+            .send()
+            .then(() => callback(null), callback);
         }
       });
 
-    return channel;
+    this.services.push(speaker);
+    return speaker;
+  }
+
+  setupInputs(television) {
+    var identifiersTLV = Buffer.alloc(0);
+    this.inputs.forEach((input, index) => {
+      const hapID = index + 1;
+      this.setupInput(input.id, input.name, hapID, television);
+
+      if (identifiersTLV.length !== 0) {
```

Comment:
How did you figure this logic out? I don't even see DisplayOrder in [hap-nodejs](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Characteristic.js).

Reply:
Yeah, see `DisplayOrderTypes` above.